cmake/CeleritasUtils.cmake

```diff
@@ -43,15 +43,180 @@ macro(celeritas_find_package_config _package)
-  set_target_properties(${target} PROPERTIES
-    LINKER_LANGUAGE CUDA
-    CUDA_SEPARABLE_COMPILATION ON
-  )
+return()
+
+#  Readd when using
+  #set_target_properties(${target} PROPERTIES
+  #  LINKER_LANGUAGE CUDA
+  #  CUDA_SEPARABLE_COMPILATION ON
+  #)
+
+  # Note: the repeat (target name, location) below is due the author of  cuda_add_library_depend
+  # not knowing how to automatically go from the target to the real file from a generator expression in add_custom_command
+  get_property(vecgeom_static_target_location TARGET VecGeom::vecgeomcuda_static PROPERTY LOCATION)
+  cuda_add_library_depend(${target} VecGeom::vecgeom VecGeom::vecgeomcuda_static ${vecgeom_static_target_location})
   target_link_libraries(${target}
     PRIVATE
+    VecGeom::vecgeom
     VecGeom::vecgeomcuda
     VecGeom::vecgeomcuda_static
   )
+  if(BUILD_SHARED_LIBS)
+  target_link_libraries(${target}_static
+    VecGeom::vecgeom
+    VecGeom::vecgeomcuda_static
+    ${PRIVATE_DEPS}
+  )
+  set_target_properties(${target}_static
+    PROPERTIES LINKER_LANGUAGE CXX
+  )
+
+  get_target_property(target_interface_include_directories ${target}
+    INTERFACE_INCLUDE_DIRECTORIES
+  )
+  set_target_properties(${target}_static PROPERTIES
+    INTERFACE_INCLUDE_DIRECTORIES "${target_interface_include_directories}")
+
+  get_target_property(target_LINK_LIBRARIES ${target}
+    LINK_LIBRARIES
+  )
+  get_target_property(target_INTERFACE_LINK_LIBRARIES ${target}
+    INTERFACE_LINK_LIBRARIES )
+  set_target_properties(${target}_static PROPERTIES
+    LINK_LIBRARIES "${target_LINK_LIBRARIES}"
+    INTERFACE_LINK_LIBRARIES "${target_INTERFACE_LINK_LIBRARIES}"
+  )
+
+  set_target_properties(${target}_static PROPERTIES
+    POSITION_INDEPENDENT_CODE ON
+  )
+endif()
+
 endfunction()
 
+
+define_property(TARGET PROPERTY CELERITAS_CUDA_LIBRARY_TYPE
+  BRIEF_DOCS "Indicate the type of cuda library (STATIC and SHARED for nvlink usage, FINAL for linking into not cuda library/executable"
+  FULL_DOCS "Indicate the type of cuda library (STATIC and SHARED for nvlink usage, FINAL for linking into not cuda library/executable"
+)
+define_property(TARGET PROPERTY CELERITAS_CUDA_FINAL_LIBRARY
+  BRIEF_DOCS "Name of the final library corresponding to this cuda library"
+  FULL_DOCS "Name of the final library corresponding to this cuda library"
+)
+define_property(TARGET PROPERTY CELERITAS_CUDA_STATIC_LIBRARY
+  BRIEF_DOCS "Name of the static library corresponding to this cuda library"
+  FULL_DOCS "Name of the static library corresponding to this cuda library"
+)
+define_property(TARGET PROPERTY CELERITAS_CUDA_MIDDLE_LIBRARY
+  BRIEF_DOCS "Name of the shared (without nvlink step) library corresponding to this cuda library"
+  FULL_DOCS "Name of the shared (without nvlink step) library corresponding to this cuda library"
+)
+
+function(celeritas_add_library target)
+  if(NOT BUILD_SHARED_LIBS OR NOT CELERITAS_USE_CUDA)
+    list(SUBLIST ARGV 1 -1 NEWARGV)
```

Comment:
I think you can omit this line and use `${ARGN}` instead of `${NEWARGV}`: that's [the standard cmake way](https://cmake.org/cmake/help/latest/command/function.html?highlight=argn#arguments) of saying all arguments past the ones listed in the function signature.